code/modules/power/supermatter/supermatter_hypermatter_state.dm

```diff
@@ -0,0 +1,195 @@
+
+
+/obj/projectile/beam/nuclear_emitter/hitscan
+	hitscan = TRUE
+	range = 10
+	tracer_type = /obj/effect/projectile/tracer/heavy_laser
+	muzzle_type = /obj/effect/projectile/muzzle/heavy_laser
+	impact_type = /obj/effect/projectile/impact/heavy_laser
+
+/obj/projectile/beam/nuclear_emitter/hitscan/Initialize(mapload)
+	. = ..()
+	color = color_matrix_rotate_hue(120)
+
+/obj/machinery/power/nuclear_emitter
+	name = "nuclear emitter"
+	desc = "A large, powerful nuclear emitter. It is capable of storing a large amount of power."
+	icon = 'icons/obj/supermatter.dmi'
+	icon_state = "nuclear_emitter"
+	base_icon_state = "nuclear_emitter"
+	circuit = /obj/item/circuitboard/machine/nuclear_emitter
+	anchored = FALSE
+	density = TRUE
+	set_dir_on_move = FALSE
+	///Amount of power stored inside
+	var/internal_energy = 0
+	///Amount required to fire a beam
+	var/needed_energy_to_emit = 1000000
+	///Is ready to fire?
+	var/ready = FALSE
+	///Is shooting a beam?
+	var/shooting = FALSE
+
+/obj/machinery/power/nuclear_emitter/examine(mob/user)
+	. = ..()
+	if(!ready)
+		. += "It is not ready to shoot."
+		. += "Current energy: [internal_energy] Watts, [needed_energy_to_emit] Watts needed to shoot."
+	else
+		. += "It is ready to shoot."
+
+/obj/machinery/power/nuclear_emitter/wrench_act(mob/living/user, obj/item/tool)
+	. = ..()
+	default_unfasten_wrench(user, tool)
+	if(anchored)
+		connect_to_network()
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/nuclear_emitter/wrench_act_secondary(mob/living/user, obj/item/tool)
+	. = ..()
+	default_change_direction_wrench(user, tool)
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/nuclear_emitter/crowbar_act(mob/living/user, obj/item/tool)
+	. = ..()
+	default_deconstruction_crowbar(tool)
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/nuclear_emitter/screwdriver_act(mob/living/user, obj/item/tool)
+	. = ..()
+	default_deconstruction_screwdriver(user, icon_state, icon_state, tool)
+	update_appearance()
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/power/nuclear_emitter/should_have_node()
+	return anchored
+
+/obj/machinery/power/nuclear_emitter/process()
+	if(!powernet || shooting)
+		return
+
+	if(!ready)
+		use_power(20000)
+		internal_energy += 20000
+
+	if(internal_energy >= needed_energy_to_emit)
+		ready = TRUE
+
+/obj/machinery/power/nuclear_emitter/attack_hand(mob/living/user, list/modifiers)
+	if(!ready || shooting)
+		return
+
+	shooting = TRUE
+	ready = FALSE
+
+	var/steps = 10
+	var/found = FALSE
+	var/turf/starting_point = get_turf(src)
+	var/obj/machinery/power/supermatter_crystal/crystal
+	while(steps > 0 || !found)
+		var/turf/next_step = get_step(starting_point, dir)
+		if(!next_step || isclosedturf(next_step))
+			shooting = FALSE
+			return
+		for(var/atom/checked_atom in next_step.contents)
+			if(istype(checked_atom, /obj/machinery/power/supermatter_crystal))
+				crystal = checked_atom
+				found = TRUE
+				break
+		starting_point = next_step
+		steps--
```

Comment:
why not make this a for loop? also this is really cludgy and I don't like it by default, why do it like this?

Reply:
i didn't have a way to find something in a specified direction, so i made this.
i could do a for loop too